builtin/logical/aws/client.go

```diff
@@ -23,91 +24,139 @@ import (
-			}
-
-			fetcher := &PluginIdentityTokenFetcher{
-				sys:      b.System(),
-				logger:   b.Logger(),
-				ns:       ns,
-				audience: config.IdentityTokenAudience,
-				ttl:      config.IdentityTokenTTL,
-			}
-
-			sessionSuffix := strconv.FormatInt(time.Now().UnixNano(), 10)
-			credsConfig.RoleSessionName = fmt.Sprintf("vault-aws-secrets-%s", sessionSuffix)
-			credsConfig.WebIdentityTokenFetcher = fetcher
-			credsConfig.RoleARN = config.RoleARN
+		if len(config.STSFallbackRegions) > 0 {
+			regions = append(regions, config.STSFallbackRegions...)
 		}
 	}
 
-	if credsConfig.Region == "" {
-		credsConfig.Region = os.Getenv("AWS_REGION")
-		if credsConfig.Region == "" {
-			credsConfig.Region = os.Getenv("AWS_DEFAULT_REGION")
-			if credsConfig.Region == "" {
-				credsConfig.Region = "us-east-1"
-			}
+	if config.IdentityTokenAudience != "" {
+		ns, err := namespace.FromContext(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get namespace from context: %w", err)
 		}
+
+		fetcher := &PluginIdentityTokenFetcher{
+			sys:      b.System(),
+			logger:   b.Logger(),
+			ns:       ns,
+			audience: config.IdentityTokenAudience,
+			ttl:      config.IdentityTokenTTL,
+		}
+
+		sessionSuffix := strconv.FormatInt(time.Now().UnixNano(), 10)
+		credsConfig.RoleSessionName = fmt.Sprintf("vault-aws-secrets-%s", sessionSuffix)
+		credsConfig.WebIdentityTokenFetcher = fetcher
+		credsConfig.RoleARN = config.RoleARN
 	}
 
-	credsConfig.HTTPClient = cleanhttp.DefaultClient()
+	if len(regions) == 0 {
+		regions = append(regions, fallbackRegion)
+	}
 
-	credsConfig.Logger = logger
+	if len(regions) != len(endpoints) {
+		// this probably can't happen, if the input was checked correctly
+		return nil, errors.New("number of regions does not match number of endpoints")
+	}
 
-	creds, err := credsConfig.GenerateCredentialChain()
-	if err != nil {
-		return nil, err
+	for i := 0; i < len(endpoints); i++ {
+		if len(regions) > i {
+			credsConfig.Region = regions[i]
+		} else {
+			credsConfig.Region = fallbackRegion
+		}
+		creds, err := credsConfig.GenerateCredentialChain()
+		if err != nil {
+			return nil, err
+		}
+		configs = append(configs, &aws.Config{
+			Credentials: creds,
+			Region:      aws.String(credsConfig.Region),
+			Endpoint:    aws.String(endpoints[i]),
+			MaxRetries:  aws.Int(maxRetries),
+			HTTPClient:  cleanhttp.DefaultClient(),
+		})
 	}
 
-	return &aws.Config{
-		Credentials: creds,
-		Region:      aws.String(credsConfig.Region),
-		Endpoint:    &endpoint,
-		HTTPClient:  cleanhttp.DefaultClient(),
-		MaxRetries:  aws.Int(maxRetries),
-	}, nil
+	return configs, nil
 }
 
 func (b *backend) nonCachedClientIAM(ctx context.Context, s logical.Storage, logger hclog.Logger) (*iam.IAM, error) {
 	awsConfig, err := b.getRootConfig(ctx, s, "iam", logger)
 	if err != nil {
 		return nil, err
 	}
-	sess, err := session.NewSession(awsConfig)
+	if len(awsConfig) != 1 {
+		return nil, errors.New("could not obtain aws config")
+	}
```

Comment:
Is this case only likely to happen when setting the new fallback fields while using IAM credentials instead of STS?

Reply:
the fields should only get appended to in the case of using STS client type, so theoretically this case should never happen